test/Algebra/Graph/Test/Label.hs

```diff
@@ -0,0 +1,147 @@
+
+semigroup :: Eq a => BinaryFn a -> a -> a -> a -> Property
+semigroup = associative
+
+monoid :: Eq a => BinaryFn a -> Identity a -> a -> a -> a -> Property
+monoid f e a b c = conjoin
+  [ associative f a b c
+  , identity f e a ]
+
+commutativeMonoid :: Eq a => BinaryFn a -> Identity a -> a -> a -> a -> Property
+commutativeMonoid f e a b c = conjoin
+  [ monoid f e a b c
+  , commutative f a b ]
+
+leftNearRing :: Eq a => Plus a -> Zero a -> Mult a -> One a -> a -> a -> a -> Property
+leftNearRing p z m o a b c = conjoin
+  [ commutativeMonoid p z a b c // "Commutative monoid"
+  , monoid m o a b c            // "Monoid"
+  , leftDistributive p m a b c
+  , annihilator m z a ]
+
+rightNearRing :: Eq a => Plus a -> Zero a -> Mult a -> One a -> a -> a -> a -> Property
+rightNearRing p z m o a b c = conjoin
+  [ commutativeMonoid p z a b c // "Commutative monoid"
+  , monoid m o a b c            // "Monoid"
+  , rightDistributive p m a b c
+  , annihilator m z a ]
+
+semiring :: Eq a => Plus a -> Zero a -> Mult a -> One a -> a -> a -> a -> Property
+semiring p z m o a b c = conjoin
+  [ commutativeMonoid p z a b c // "Commutative monoid"
+  , monoid m o a b c            // "Monoid"
+  , distributive p m a b c
+  , annihilator m z a ]
+
+dioid :: Eq a => Plus a -> Zero a -> Mult a -> One a -> a -> a -> a -> Property
+dioid p z m o a b c = conjoin
+  [ semiring p z m o a b c
+  , idempotent p a ]
+
+starSemiring :: Eq a => Plus a -> Zero a -> Mult a -> One a -> Star a -> a -> a -> a -> Property
+starSemiring p z m o s a b c = conjoin
+  [ semiring p z m o a b c
+  , closure p m o s a ]
+
+testLeftNearRing :: (Eq a, Semiring a) => a -> a -> a -> Property
+testLeftNearRing = leftNearRing (<+>) zero (<.>) one
+
+testSemiring :: (Eq a, Semiring a) => a -> a -> a -> Property
+testSemiring = semiring (<+>) zero (<.>) one
+
+testDioid :: (Eq a, Dioid a) => a -> a -> a -> Property
+testDioid = dioid (<+>) zero (<.>) one
+
+testStarSemiring :: (Eq a, StarSemiring a) => a -> a -> a -> Property
+testStarSemiring = starSemiring (<+>) zero (<.>) one star
+
+testLabel :: IO ()
+testLabel = do
+    putStrLn "\n============ Any ============"
+    test "Semiring"     $ \(a :: Any) b c -> testSemiring a b c
+    test "StarSemiring" $ \(a :: Any) b c -> testStarSemiring a b c
+    test "Dioid"        $ \(a :: Any) b c -> testDioid a b c
+
+    putStrLn "\n============ Distance ============"
+    test "Semiring"     $ \(a :: Distance Int) b c -> testSemiring a b c
+    test "StarSemiring" $ \(a :: Distance Int) b c -> testStarSemiring a b c
+    test "Dioid"        $ \(a :: Distance Int) b c -> testDioid a b c
+
+    putStrLn "\n============ Capacity ============"
+    test "Semiring"     $ \(a :: Capacity Int) b c -> testSemiring a b c
+    test "StarSemiring" $ \(a :: Capacity Int) b c -> testStarSemiring a b c
+    test "Dioid"        $ \(a :: Capacity Int) b c -> testDioid a b c
+
+    putStrLn "\n============ Minimum ============"
```

Comment:
You are actually testing `Minimum (Path Int)` below, not `Minimum`.

I think we should add tests for all types of labels we define in `Algebra.Graph.Label`, perhaps, also adding some combinations like `Minimum (Path Int)`.

Reply:
> perhaps, also adding some combinations like Minimum (Path Int).
Perhaps you meant something else? `Minimum (Path Int)` already exists.

Reply:
I meant similar to `Minimum Path` that you are already testing.

Reply:
I will do it in another PR. Some types (`Optimum`) need not satisfy `Semiring` yet can be used. I need to come up with a proper/better reasoning.